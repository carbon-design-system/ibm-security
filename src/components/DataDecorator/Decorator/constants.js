/**
 * @file Decorator constants.
 * @copyright IBM Security 2019
 */

import { getComponentNamespace } from '../../../globals/namespace';

const namespace = getComponentNamespace('decorator');

const unknown = `M8 3.5c2.5 0 4.5 2 4.5 4.5s-2 4.5-4.5 4.5-4.5-2-4.5-4.5 2-4.5 4.5-4.5M8 2C4.7 2 2 4.7 2 8s2.7 6 6 6 6-2.7 6-6-2.7-6-6-6z`;

const critical = `M14 2v12H2z`;

const high = `M8 2l7 11H1z`;

const medium = `M 8.00,1.00
           C 8.00,1.00 15.00,8.00 15.00,8.00
             15.00,8.00 8.00,15.00 8.00,15.00
             8.00,15.00 1.00,8.00 1.00,8.00
             1.00,8.00 8.00,1.00 8.00,1.00 Z`;

const low = `M 14.00,2.00
           C 14.00,2.00 14.00,14.00 14.00,14.00
             14.00,14.00 2.00,14.00 2.00,14.00
             2.00,14.00 2.00,2.00 2.00,2.00
             2.00,2.00 14.00,2.00 14.00,2.00 Z`;

const icons = {
  critical,
  high,
  low,
  medium,
  unknown,
};

const getDecoratorProps = (score, scoreThresholds, active) => {
  let path = icons.unknown;
  let classes = `${namespace}--unknown`;

  if (score < scoreThresholds[1] && score >= scoreThresholds[0]) {
    path = icons.low;
    classes = `${namespace}--low`;
  } else if (score < scoreThresholds[2] && score >= scoreThresholds[1]) {
    path = icons.medium;
    classes = `${namespace}--medium`;
  } else if (score < scoreThresholds[3] && score >= scoreThresholds[2]) {
    path = icons.high;
    classes = `${namespace}--high${active ? '--active' : ''}`;
  } else if (score >= scoreThresholds[3]) {
    path = icons.critical;
    classes = `${namespace}--critical`;
  }

  return { path, classes };
};

export { getDecoratorProps, icons, namespace };
