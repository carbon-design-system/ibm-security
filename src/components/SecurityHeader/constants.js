/**
 * @file Header constants.
 * @copyright IBM Security 2018
 */

import PropTypes from 'prop-types';

import { getComponentNamespace } from '../../globals/namespace';

const defaultProps = {
  className: null,
  onNotificationClear: notificationId => notificationId,
  notifications: [],
  profile: null,
  showEditProfile: true,
  showNotifications: true,
  totalNotifications: 0,
};

const namespace = getComponentNamespace('header');

const propTypes = {
  /** @type {string} Extra classes. */
  className: PropTypes.string,

  /** @type {Object.<Object, *>} An object list of labels. */
  labels: PropTypes.shape({
    /** @type {Object.<string, string>} An object list of brand labels. */
    brand: PropTypes.shape({
      /** @type {string} The company. */
      company: PropTypes.string.isRequired,

      /** The domain. */
      domain: PropTypes.string,

      /** @type {string} The product name. */
      product: PropTypes.string.isRequired,
    }).isRequired,

    /** @type {Object.<string, string>} An object list of notifications labels. */
    notifications: PropTypes.shape({
      /** @type {string} The notification button label. */
      button: PropTypes.string.isRequired,

      /** @type {string} The notification clear button label. */
      clear: PropTypes.string.isRequired,

      /** @type {string} The notification button clear all label. */
      clear_all: PropTypes.string.isRequired,

      /** @type {string} The notifications link label. */
      link: PropTypes.string.isRequired,

      /** @type {string} The notifications success label. */
      success: PropTypes.string.isRequired,

      /** @type {string} The notifications today label. */
      today: PropTypes.string.isRequired,

      /** @type {string} The notifications title label. */
      title: PropTypes.string.isRequired,

      /** @type {string} The notification 'via' label. */
      via: PropTypes.string.isRequired,

      /** @type {string} The svg title text for the notification "preferences" icon link. */
      preferences: PropTypes.string,
    }).isRequired,

    /** @type {Object.<string, string>} An object list of profile labels. */
    profile: PropTypes.shape({
      /** @type {string} The account label. */
      link: PropTypes.string,

      /** @type {string} Registration label. */
      registration: PropTypes.string.isRequired,

      /** @type {string} Logging in label. */
      sign_in: PropTypes.string.isRequired,

      /** @type {string} Logging out label. */
      sign_out: PropTypes.string.isRequired,

      /** @type {string} Edit profile label. */
      edit_profile: PropTypes.string,

      /** @type {string} Account label. */
      account: PropTypes.string,
    }).isRequired,
  }).isRequired,

  /** @type {Array<Object.*>} An array list of accounts. */
  accounts: PropTypes.arrayOf(
    /** @type {Object.<string, string>} An object list of accounts. */
    PropTypes.shape({
      /** @type {string} The account id. */
      id: PropTypes.string.isRequired,

      /** @type {string} The account name. */
      name: PropTypes.string.isRequired,
    })
  ),

  /** @type {Function} The callback when an account is clicked. */
  onAccountClick: PropTypes.func,

  /** @type {Array<Object.*>} An array list of notifications. */
  notifications: PropTypes.arrayOf(
    /** @type {Object.<string, string>} An object list of notifications. */
    PropTypes.shape({
      /** @type {string} The notification ID. */
      id: PropTypes.string.isRequired,

      /** @type {string} The notification date. */
      datetime: PropTypes.string.isRequired,

      /** @type {string} The notification description. */
      description: PropTypes.string.isRequired,

      /** @type {string} The notification URL. */
      href: PropTypes.string,

      /** @type {string} The notification label. */
      label: PropTypes.string.isRequired,

      /** @type {string} The product name. */
      product: PropTypes.string.isRequired,
    }).isRequired
  ),

  /** @type {Object.<string, string>} An object list of URLs. */
  links: PropTypes.shape({
    /** @type {string} The product URL. */
    product: PropTypes.string.isRequired,

    /** @type {string} The profile URL. */
    profile: PropTypes.string.isRequired,

    /** @type {string} The registration URL. */
    registration: PropTypes.string.isRequired,

    /** @type {string} The URL to sign in. */
    sign_in: PropTypes.string.isRequired,

    /** @type {string} The URL to sign out. */
    sign_out: PropTypes.string.isRequired,

    /** @type {string} The URL to edit profile. */
    edit_profile: PropTypes.string,

    /** @type {string} The preferences href in the notifications popover. */
    notifications_preferences: PropTypes.string,

    /** @type {string} The 'view all' href in the notifications popover. */
    notifications_view_all: PropTypes.string,
  }).isRequired,

  /** @type {Function} The callback when a notification is cleared. */
  onNotificationClear: PropTypes.func,

  /** @type {Object.<Object, *>} An object list of profile information. */
  profile: PropTypes.shape({
    /** @type {string} The profile image URL. */
    image_url: PropTypes.string,

    /** @type {string} The profile email. */
    email: PropTypes.string,

    /** @type {accounts} The current user account. */
    account: PropTypes.shape({
      /** @type {string} The account id. */
      id: PropTypes.string.isRequired,

      /** @type {string} The account name. */
      name: PropTypes.string.isRequired,
    }),

    /** @type {Object.<string, string>} An object list of profile names. */
    name: PropTypes.shape({
      /** @type {string} The profile first name. */
      first_name: PropTypes.string.isRequired,

      /** @type {string} The profile surname. */
      surname: PropTypes.string.isRequired,
    }).isRequired,
  }),

  /** @type {boolean} Whether or not to show the Edit Profile link. */
  showEditProfile: PropTypes.bool,

  /** @type {boolean} Whether or not to show the notifications bell. */
  showNotifications: PropTypes.bool,

  /** @type {number} The total number of unread notifications of the current user. */
  totalNotifications: PropTypes.number.isRequired,
};

export { defaultProps, namespace, propTypes };
