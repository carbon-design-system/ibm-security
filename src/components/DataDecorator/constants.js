/**
 * @file DataDecorator constants.
 * @copyright IBM Security 2019-2020
 */

import { getComponentNamespace } from '../../globals/namespace';
import { icons } from './Decorator/constants';

const namespace = getComponentNamespace('decorator');

export { icons, namespace };
