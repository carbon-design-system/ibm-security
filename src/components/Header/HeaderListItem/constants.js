/**
 * @file Header list item constants.
 * @copyright IBM Security 2018
 */

import PropTypes from 'prop-types';

import { appendComponentNamespace } from '../../../globals/namespace';
import { namespace as headerNamespace } from '../constants';

const defaultProps = {
  className: '',
};

const namespace = appendComponentNamespace(headerNamespace, 'list__item');

const propTypes = {
  /** @type {node} List item children. */
  children: PropTypes.node.isRequired,

  /** @type {string} Extra classes. */
  className: PropTypes.string,
};

export { defaultProps, namespace, propTypes };
