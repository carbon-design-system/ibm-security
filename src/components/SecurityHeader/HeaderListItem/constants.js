/**
 * @file Header list item constants.
 * @copyright IBM Security 2018
 */

import PropTypes from 'prop-types';

import { appendComponentNamespace } from '../../../globals/namespace';
import { namespace as headerNamespace } from '../constants';

const defaultProps = {
  className: '',
  hasPopup: false,
  isExpanded: false,
};

const namespace = appendComponentNamespace(headerNamespace, 'list__item');

const propTypes = {
  /** @type {boolean|HTMLElement} List item children. */
  children: PropTypes.oneOfType([PropTypes.bool, PropTypes.element]).isRequired,

  /** @type {string} Extra classes. */
  className: PropTypes.string,

  /** @type {boolean} Whether it has a popover or not. */
  hasPopup: PropTypes.bool,

  /** @type {boolean} Whether it is expanded or not. */
  isExpanded: PropTypes.bool,
};

export { defaultProps, namespace, propTypes };
