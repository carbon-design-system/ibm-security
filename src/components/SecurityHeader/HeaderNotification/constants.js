/**
 * @file Header notification constants.
 * @copyright IBM Security 2018
 */

import PropTypes from 'prop-types';

import { appendComponentNamespace } from '../../../globals/namespace';
import { namespace as headerNamespace } from '../constants';

const namespace = appendComponentNamespace(headerNamespace, 'notification');

const propTypes = {
  /** @type {string} Clear button label. */
  clearButtonLabel: PropTypes.string.isRequired,

  /** @type {string} Date. */
  dateTime: PropTypes.string.isRequired,

  /** @type {string} Description. */
  description: PropTypes.string.isRequired,

  /** @type {Function} Clear button callback. */
  onClearButtonClick: PropTypes.func.isRequired,

  /** @type {string} Product name. */
  product: PropTypes.string.isRequired,

  /** @type {string} Time label. */
  timeLabel: PropTypes.string.isRequired,

  /** @type {string} 'Via' label. */
  viaLabel: PropTypes.string.isRequired,

  /** @type {boolean} Has the notification been acknowledged. */
  acknowledged: PropTypes.bool,
};

export { namespace, propTypes };
