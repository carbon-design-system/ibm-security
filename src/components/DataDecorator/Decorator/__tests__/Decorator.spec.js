/**
 * @file Decorator tests.
 * @copyright IBM Security 2019 - 2021
 */

import userEvent from '@testing-library/user-event';
import { render } from '@testing-library/react';

import React from 'react';

import { Decorator } from '../../../..';

import { namespace, icons } from '../constants';

describe('Decorator', () => {
  test('has no accessibility violations when rendered as a button', async () => {
    const { container } = render(
      <Decorator type="IP" value="10.0.0.0" score={0} />
    );

    await expect(container).toBeAccessible('Decorator as a button');
    await expect(container).toHaveNoAxeViolations();
  });

  test('has no accessibility violations when rendered as a link', async () => {
    const { container } = render(
      <Decorator type="IP" value="10.0.0.0" score={0} href="#" />
    );

    await expect(container).toBeAccessible('Decorator as a link');
    await expect(container).toHaveNoAxeViolations();
  });

  test('has no accessibility violations when inert', async () => {
    const { container } = render(
      <Decorator type="IP" value="10.0.0.0" score={0} invert />
    );

    await expect(container).toBeAccessible('Decorator as inert');
    await expect(container).toHaveNoAxeViolations();
  });

  test('should apply a score', () => {
    const { queryByTitle } = render(
      <Decorator type="IP" value="10.0.0.0" score={5} />
    );
    // Expect svg's `title` text, which is in the DOM but not necessarily visible:
    expect(queryByTitle(/5/i)).toBeInTheDocument();
  });

  test('should apply a score description', () => {
    const { queryByTitle } = render(
      <Decorator
        type="IP"
        value="10.0.0.0"
        score={5}
        scoreDescription={score => `Test score ${score}`}
      />
    );
    // Expect svg's `title` text, which is in the DOM but not necessarily visible:
    expect(queryByTitle(/Test score 5/i)).toBeInTheDocument();
  });

  test('should apply a value', () => {
    const { queryByText } = render(<Decorator type="IP" value="10.0.0.0" />);
    expect(queryByText(/10.0.0.0/i)).toBeVisible();
  });

  test('should apply a `title` attribute to the `value` element via the `title` prop', () => {
    const { queryByTitle } = render(
      <Decorator type="IP" value="10.0.0.0" title="test title" />
    );
    expect(queryByTitle(/test title/i)).toBeInTheDocument();
  });

  test('should accept a custom class', () => {
    const { container } = render(
      <Decorator type="IP" value="10.0.0.0" className="custom-class" />
    );
    expect(container.firstElementChild).toHaveClass('custom-class');
  });

  test('should invoke click mock when decorator is clicked', () => {
    const onClickMock = jest.fn();
    const { getByText } = render(
      <Decorator type="IP" value="10.0.0.0" onClick={onClickMock} />
    );

    userEvent.click(getByText(/10.0.0.0/i).closest('button'));
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  test('should apply an href', () => {
    const { getByText } = render(
      <Decorator type="IP" value="10.0.0.0" href="#" />
    );
    expect(getByText(/10.0.0.0/i).closest('a')).toHaveAttribute('href', '#');
  });

  test('should render as a `span` by default', () => {
    const { container } = render(<Decorator type="IP" value="10.0.0.0" />);
    expect(container.firstElementChild.nodeName).toBe('SPAN');
  });

  test('should render as an `a` element when an `href` is provided', () => {
    const { container } = render(
      <Decorator type="IP" value="10.0.0.0" href="#" />
    );
    expect(container.firstElementChild.nodeName).toBe('A');
  });

  test('should render as a `button` when `onClick` is provided', () => {
    const { container } = render(
      <Decorator type="IP" value="10.0.0.0" onClick={() => {}} />
    );
    expect(container.firstElementChild.nodeName).toBe('BUTTON');
  });

  test('should apply correct active class when `active` is set `true`', () => {
    const { container } = render(
      <Decorator type="IP" value="10.0.0.0" active />
    );
    expect(container.firstElementChild).toHaveClass(`${namespace}--active`);
  });

  test('should apply correct inline class when `inline` is set `true`', () => {
    const { container } = render(
      <Decorator type="IP" value="10.0.0.0" inline />
    );
    expect(container.firstElementChild).toHaveClass(`${namespace}--inline`);
  });

  test('should use the value of the decorator to the panel title when mid-line truncation is required', () => {
    const { queryAllByText } = render(
      <Decorator
        type="IP"
        value="0f3deda483df5e5f8043ea20297d243b"
        score={0}
        midLineTruncation={{ enabled: true, maxLength: 20, front: 12, back: 4 }}
      />
    );
    // the MD5 hash -- the data decorator.
    expect(queryAllByText(/0f3deda483dfâ€¦243b/i).length === 1);
  });
});

Object.keys(icons).forEach(icon => {
  // Capitalize first character of icon name
  // to correctly call component name (like `Decorator.Low`, etc.)
  const formattedName = icon.charAt(0).toUpperCase() + icon.slice(1);
  const className = `${namespace}__icon--${icon}`;

  const Component = Decorator[formattedName];

  describe(`Decorator.${formattedName}`, () => {
    test('has no accessibility violations', async () => {
      const { container } = render(
        <Component description={`${formattedName} severity`} />
      );

      await expect(container).toBeAccessible(`Decorator.${formattedName}`);
      await expect(container).toHaveNoAxeViolations();
    });

    test('should apply accessible `aria-label` with `description` prop', () => {
      const { getByLabelText } = render(
        <Component description={`${formattedName} severity`} />
      );
      expect(getByLabelText(`${formattedName} severity`)).toBeVisible();
    });

    test('should add a custom class', () => {
      render(
        <Component
          className="custom-class"
          description={`${formattedName} severity`}
        />
      );
      expect(document.querySelector(`.${className}`)).toHaveClass(
        'custom-class'
      );
    });

    test('should pass through extra props via spread attribute', () => {
      const { queryByTestId } = render(
        <Component
          data-testid="test-id"
          description={`${formattedName} severity`}
        />
      );
      expect(queryByTestId('test-id')).toBeVisible();
    });

    test('should set `height` and `width` with `size` prop', () => {
      render(<Component size={12} description={`${formattedName} severity`} />);
      expect(document.querySelector(`.${className}`)).toHaveAttribute(
        'height',
        '12'
      );
      expect(document.querySelector(`.${className}`)).toHaveAttribute(
        'width',
        '12'
      );
    });
  });
});
