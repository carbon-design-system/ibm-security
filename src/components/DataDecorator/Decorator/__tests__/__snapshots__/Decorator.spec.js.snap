// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decorator renders correctly 1`] = `
<Link
  className="security--decorator security--decorator--low security--decorator--link"
  href="#"
  tabIndex={0}
>
  <span
    className="security--decorator__icon"
  >
    <Icon
      className=""
      fillRule="evenodd"
      height={null}
      path="M 8.00,3.00
           C 8.00,3.00 14.00,12.00 14.00,12.00
             14.00,12.00 2.00,12.00 2.00,12.00
             2.00,12.00 8.00,3.00 8.00,3.00 Z"
      renderIcon={null}
      size={16}
      viewBox="0 0 16 16"
      width={null}
    />
  </span>
  <span
    className="security--decorator__type"
  >
    IP
  </span>
  <span
    className="security--decorator__value"
    title=""
  >
    10.0.0.0
  </span>
</Link>
`;

exports[`Decorator renders the \`inert\` variation 1`] = `
<Link
  className="security--decorator security--decorator--low security--decorator--link security--decorator--inert"
  href="#"
  tabIndex={0}
>
  <span
    className="security--decorator__icon"
  >
    <Icon
      className=""
      fillRule="evenodd"
      height={null}
      path="M 8.00,3.00
           C 8.00,3.00 14.00,12.00 14.00,12.00
             14.00,12.00 2.00,12.00 2.00,12.00
             2.00,12.00 8.00,3.00 8.00,3.00 Z"
      renderIcon={null}
      size={16}
      viewBox="0 0 16 16"
      width={null}
    />
  </span>
  <span
    className="security--decorator__type"
  >
    IP
  </span>
  <span
    className="security--decorator__value"
    title=""
  >
    10.0.0.0
  </span>
</Link>
`;

exports[`Decorator renders the HTML of the node's subtree 1`] = `
<a
  class="bx--link security--decorator security--decorator--low security--decorator--link"
  href="#"
  tabindex="0"
>
  <span
    class="security--decorator__icon"
  >
    <svg
      aria-hidden="true"
      class="security--icon"
      fill-rule="evenodd"
      focusable="false"
      height="16"
      preserveAspectRatio="xMidYMid meet"
      style="will-change:transform"
      viewBox="0 0 16 16"
      width="16"
    >
      <path
        d="M 8.00,3.00
           C 8.00,3.00 14.00,12.00 14.00,12.00
             14.00,12.00 2.00,12.00 2.00,12.00
             2.00,12.00 8.00,3.00 8.00,3.00 Z"
      />
    </svg>
  </span>
  <span
    class="security--decorator__type"
  >
    IP
  </span>
  <span
    class="security--decorator__value"
    title=""
  >
    10.0.0.0
  </span>
</a>
`;
